Bluetooth: Mesh: Time models documentation (#2972)
nrfconnect/sdk-nrf: include/bluetooth/mesh/time_srv.rst

.. _bt_mesh_time_srv_readme:

Time Server
###########

The Time Server model allows mesh nodes to synchronize the current time and date by publishing and subscribing to Time Status messages.

The Time Server model relies on an external clock source to provide the current time and date, but takes care of distributing this information through the network.

The Time Server model adds the following new model instances in the composition data:

* Time Server
* Time Setup Server

Operation
=========

> The Time Server model builds an ad-hoc time synchronization hierarchy through the mesh network, that propagates the current wall clock time as TAI timestamps.

Reviewer: ```suggestion
The Time Server model builds an ad-hoc time synchronization hierarchy through the mesh network, that propagates the current wall-clock time as TAI timestamps.
```
Either wall-clock time or wall time :)

To ensure consistent synchronization, the Time Server model should be configured to publish its Time Status with regular intervals using a :ref:`zephyr:bluetooth_mesh_models_cfg_cli` model.

..note::

   To ensure a high level of accuracy, Time Status messages are always transmitted one hop at a time.
   This means that every mesh node that needs access to the common time reference must be within radio range of at least one other node with the Time Server model.

>    Time Server nodes that have no neighboring Time Servers need to get this information some other way.

Reviewer: ```suggestion
   Time Server nodes that have no neighboring Time Servers need to get this information in some other way.
```

   The Time Authority will publish Time Status messages, but will not receive them.

Time Relay
   A Time Server that both publishes and receives Time Status messages.
   The Time Relay relies on other mesh nodes to tell time.

Time Client
   A Time Server that receives Time Status messages, but does not relay them.
   The Time Client relies on other mesh nodes to tell time.

Whenever a Time Server model that is not a Time Authority receives a Time Status message, it will update its internal clock to match the new Time Status - if it improves its clock uncertainty.

The Time Authority role should not be confused with the Authority state, which denotes whether the device is actively monitoring a reliable external time source.

Clock uncertainty
*****************

The Time Server's time synchronization mechanism is not perfect, and every timestamp it provides has some uncertainty.
The clock uncertainty is measured in milliseconds, and can come from various sources, like inaccuracies in the original clock source, internal clock drift or uncertainty in the mesh stack itself.

> The Time Servers inherit the uncertainty of the node it got its Time Status from, and the clock uncertainty of each mesh node increases for every hop the Time Status had to make from the Time Authority.
> 

Reviewer: ```suggestion
The Time Server inherits the uncertainty of the node it got its Time Status from, and the clock uncertainty of each mesh node increases for every hop the Time Status had to make from the Time Authority.

```

The expected uncertainty in the mesh stack comes from the fluctuation in time to encrypt a message and broadcast it, and can be configured through the :option:`CONFIG_BT_MESH_TIME_MESH_HOP_UNCERTAINTY`.

In addition to the uncertainty added by the mesh stack itself, the internal clock of each device will increase the uncertainty of the timestamp over time.
As crystal hardware is not perfectly accurate, it will gradually drift away from the correct time.
The potential range of this clock drift is generally a known property of the onboard hardware, even if the actual drift fluctuates.
As it's impossible to measure the actual clock drift at any given moment, the Time Server will instead gradually increase the uncertainty of its timestamp based on the known clock accuracy.

The Time Server's notion of the local clock accuracy can be configured with :option:`CONFIG_BT_MESH_TIME_SRV_CLOCK_ACCURACY`.
By default, it uses the configured kernel clock control accuracy.

Leap seconds and time zones
***************************

> As described in the section on :ref:`bt_mesh_time_tai_readme`, the Time Server measures time as TAI seconds, and adds the UTC leap second offset and local time zone as additional information.

Reviewer 2: :ref:`bt_mesh_time_tai` ?

This function can be used in combination with the three others to determine the accuracy of a provided timestamp.

.. note::
   All time and uncertainty conversion is based on the Time Server's current data, and assumes that no corrections are made between the call and the provided timestamp.
   Timestamps that are weeks or months into the future may have an uncertainty of several hours, due to clock drift.
   The application can subscribe to changes in the Time Server state through the :cpp:type:`bt_mesh_time_srv_cb` callback structure.

   Any time zone or UTC delta changes are taken into account.

Configuration
=============

The clock uncertainty of the Time Server model can be configured with the following configuration options:

* :option:`CONFIG_BT_MESH_TIME_MESH_HOP_UNCERTAINTY`: The amount of uncertainty introduced in the mesh stack through sending a single message, in milliseconds.
* :option:`CONFIG_BT_MESH_TIME_SRV_CLOCK_ACCURACY`: The largest possible clock drift introduced by the kernel clock's hardware, in parts per million.

States
======

The Time Server model contains the following states:

TAI time: :cpp:type:`bt_mesh_time_tai`

>    The TAI time is a composite state, with members ``sec`` and an 8 bit ``subsec``.

Reviewer: ```suggestion
   The TAI time is a composite state, with members ``sec`` and an 8-bit ``subsec``.
```

   If the current time is known, the TAI time changes continuously.

Uncertainty: ``uint64_t``
   Current clock uncertainty in milliseconds.
   Without new data, clock uncertainty increases gradually due to clock drift.

UTC delta: ``int16_t``
   Number of seconds between the TAI and UTC timestamps due to UTC leap seconds.

Time zone offset: ``int16_t``

>    Local time zone offset in 15 minute increments.

Reviewer: ```suggestion
   Local time zone offset in 15-minute increments.
```
